cdap-common/bin/functions.sh

```
@@ -493,11 +491,34 @@ cdap_set_spark() {
         return 1
       fi
       SPARK_HOME=$(echo -e "${SPARK_VAR_OUT}" | grep ^SPARK_HOME= | cut -d= -f2)
-      export SPARK_HOME
-      return 0
     fi
-    return 1
   fi
+
+  export SPARK_HOME
+
+  # Find environment variables setup via spark-env.sh
+  SPARK_ENV_SH=${SPARK_HOME}/conf/spark-env.sh
+  if [[ -f ${SPARK_ENV_SH} ]]; then
+    if [[ -z ${SPARK_ENV_PATTERN} ]]; then
+      # By default, only get env from spark-env.sh that starts with PY or SPARK
+      SPARK_ENV_PATTERN="^(PY|SPARK)"
+    fi
+    SPARK_ENV=$(source ${SPARK_ENV_SH}; env | grep -E "${SPARK_ENV_PATTERN}")
+    __ret=$?
+    if [[ ${__ret} -eq 0 ]]; then
+      local readonly __env_out=$(mktemp)
+      echo "${SPARK_ENV}" | while read -r line; do
+        # Prefix the env variable with _SPARK_ to avoid conflicts
+        echo "export _SPARK_${line%%=*}=${line#*=}" >> ${__env_out}
+      done
+      source ${__env_out}
+      rm ${__env_out}
+    else
+      logecho "[WARN] Fail to source ${SPARK_ENV_SH} to setup environment variables for Spark"
+    fi
+  fi
```

Is all of this configuration optional? Should we be returning 1 on any particular failure cases?

---

It is optional for so long, even spark itself is optional

---

Yes, but we return 1 when things don't work out and let the caller decide what to do with it. If the caller wants to throw and error, it should be able to do that. If we mask any errors, that can't be done. You can see examples of this in `cdap_set_hive` where it's fatal if the Hive execution engine is set to Spark, yet Spark cannot be found, as Explore won't work, but it's optional in `cdap_start_java` and `cdap_run_class` functions.

---

returning 1 from this method will cause a message being logged, saying `Could not determine SPARK_HOME! Spark support unavailable!`, even Spark is available, just that we are not able to source the `spark-env.sh`, which can have Spark programs run perfectly fine without it (depending on the cluster configuration, and CDAP has been not including it so far). That's why I just log a warn here without having this method return 1.